pkg/entities/record.go

```diff
@@ -101,89 +106,281 @@ func (d *dataRecord) PrepareRecord() (uint16, error) {
+				if err != nil {
+					return nil, fmt.Errorf("Error in decoding val to uint16: %v", err)
+				}
+				return v, nil
+			case Unsigned32:
+				var v uint32
+				err := util.Decode(value, binary.BigEndian, &v)
+				if err != nil {
+					return nil, fmt.Errorf("Error in decoding val to uint32: %v", err)
+				}
+				return v, nil
+			case Unsigned64:
+				var v uint64
+				err := util.Decode(value, binary.BigEndian, &v)
+				if err != nil {
+					return nil, fmt.Errorf("Error in decoding val to uint64: %v", err)
+				}
+				return v, nil
+			case Signed8:
+				var v int8
+				err := util.Decode(value, binary.BigEndian, &v)
+				if err != nil {
+					return nil, fmt.Errorf("Error in decoding val to int8: %v", err)
+				}
+				return v, nil
+			case Signed16:
+				var v int16
+				err := util.Decode(value, binary.BigEndian, &v)
+				if err != nil {
+					return nil, fmt.Errorf("Error in decoding val to int16: %v", err)
+				}
+				return v, nil
+			case Signed32:
+				var v int32
+				err := util.Decode(value, binary.BigEndian, &v)
+				if err != nil {
+					return nil, fmt.Errorf("Error in decoding val to int32: %v", err)
+				}
+				return v, nil
+			case Signed64:
+				var v int64
+				err := util.Decode(value, binary.BigEndian, &v)
+				if err != nil {
+					return nil, fmt.Errorf("Error in decoding val to int64: %v", err)
+				}
+				return v, nil
+			case Float32:
+				var v float32
+				err := util.Decode(value, binary.BigEndian, &v)
+				if err != nil {
+					return nil, fmt.Errorf("Error in decoding val to float32: %v", err)
+				}
+				return v, nil
+			case Float64:
+				var v float64
+				err := util.Decode(value, binary.BigEndian, &v)
+				if err != nil {
+					return nil, fmt.Errorf("Error in decoding val to float64: %v", err)
+				}
+				return v, nil
+			case Boolean:
+				var v int
+				err := util.Decode(value, binary.BigEndian, &v)
+				if err != nil {
+					return nil, fmt.Errorf("Error in decoding val to boolean: %v", err)
+				}
+				if v == 1 {
+					return true, nil
+				} else {
+					return false, nil
+				}
+			case DateTimeSeconds, DateTimeMilliseconds:
+				var v uint64
+				err := util.Decode(value, binary.BigEndian, &v)
+				if err != nil {
+					return nil, fmt.Errorf("Error in decoding val to uint64: %v", err)
+				}
+				return v, nil
+			case DateTimeMicroseconds, DateTimeNanoseconds:
+				return nil, fmt.Errorf("This API does not support micro and nano seconds types yet")
+			case MacAddress, Ipv4Address, Ipv6Address:
+				return value.Bytes(), nil
+			case String:
+				return value.String(), nil
+			default:
+				return nil, fmt.Errorf("API supports only valid information elements with datatypes given in RFC7011")
+			}
+		}
+	}
+	return nil, fmt.Errorf("API supports only valid information elements with datatypes given in RFC7011")
+}
+
+// encodeToIEDataType is to encode to specific type to the buff
+func (d *dataRecord) encodeToIEDataType(dataType IEDataType, val interface{}) (interface{}, error) {
```

Comment:
Same here; it can be utility function... d.buff can be sent as argument. Seems more like a generic function rather than being part of dataRecord structure.